tests/test_taar_locale.py

```diff
@@ -35,142 +36,481 @@
+    return df
 
-    # Store the data in the mocked bucket.
-    conn.Object(
-        taar_utils.AMO_DUMP_BUCKET, key=taar_utils.AMO_CURATED_WHITELIST_KEY
-    ).put(Body=json.dumps(FAKE_AMO_DUMP))
 
-    multi_locales_df.createOrReplaceTempView("clients_daily")
+@pytest.fixture
+def short_client_df(client_addons_df):
+    """A condensed version of the `clients_daily` data."""
+    df = client_addons_df.where("locale in ('en-US', 'de')")
+    df.createOrReplaceTempView("clients_daily")
+    return df
+
+
+@pytest.fixture
+def multi_day_client_df(client_addons_df):
+    """A single-locale version of the `clients_daily` data."""
+    df = (
+        client_addons_df.where("locale = 'en-US'")
+        .withColumn("client_id", substring_index("client_id", "_", -1))
+        .where(substring_index("client_id", "-", -1).isin([1, 2, 3, 4, 5]))
+    )
+    df.createOrReplaceTempView("clients_daily")
+    return df
+
+
+@pytest.fixture
+def mock_sql_rand(spark):
+    """Patch Spark SQL `RAND` to return a constant value.
+
+    This will ensure deterministic selection of the add-on subsets, assuming
+    that sorting maintains the ordering of `active_addons`.
+    """
+
+    def mock_rand():
+        return 1
+
+    spark.udf.register("rand", mock_rand, IntegerType())
+    yield
+    # Afterwards, reset all temporary catalog items to undo the patching.
+    spark.catalog._reset()
+
+
+@pytest.fixture()
+def mock_rlaplace(monkeypatch):
+    """Patch the Laplace noise generation to make results deterministic."""
+
+    def mock_noise(scale, size):
+        """Return `size` copies of the `scale` parameter."""
+        return np_repeat(scale, size)
+
+    monkeypatch.setattr(taar_locale, "rlaplace", mock_noise)
+
+
+@pytest.fixture()
+def mock_amo_whitelist(monkeypatch, short_whitelist):
+    """Patch the whitelist loader to just use the one defined here."""
+    monkeypatch.setattr(
+        taar_locale, "load_amo_curated_whitelist", lambda: short_whitelist
+    )
+
+
+@pytest.fixture()
+def mock_today(monkeypatch):
+    """Patch `date.today()` to 2019-01-15, the latest date in the dataset."""
+
+    class MockDate:
+        @classmethod
+        def today(cls):
+            return date(2019, 1, 15)
+
+    monkeypatch.setattr(taar_locale, "date", MockDate)
+
+
+def spark_df_to_records(df):
+    """Dump a Spark DF as a list of tuples representing rows ('records')."""
+    return [tuple(r) for r in df.collect()]
+
+
+def pandas_df_to_records(df):
+    """Convert a Pandas DF to a list of tuples representing rows ('records')."""
+    return df.to_records(index=False).tolist()
+
+
+def same_rows(rows_list_1, rows_list_2):
+    """Compare DF rows represented as lists of tuples ('records').
+
+    Checks that the lists contain the same tuples (possibly in different orders).
+    """
+    return sorted(rows_list_1) == sorted(rows_list_2)
+
+
+# =========================================================================== #
+
```

Comment:
Please include an additional test that locale/add-on output weights sum to 1.0 at output.

Similar expected output as is used in `test_get_top_addons_by_locale(addon_count_neg):` could be used, but I'd like to see the additional sanity check that our relative weights are comparable across locales after the noisification, scaling, and removal operations are all performed.